### PR TITLE
Custom op shape inference API

### PR DESCRIPTION
**Pull request by RandySheriffH**
Add c/cxx API to allow custom ops do shape  inference.

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_lite_custom_op.h [hunk 1]
```diff
@@ -928,6 +930,113 @@ struct OrtLiteCustomFunc : public OrtLiteCustomOp {
...
+        } else {
+          shape.emplace_back(integer_shape[ith]);
+        }
+      }
+      input_shapes_.push_back(std::move(shape));
+      type_shape_info.release();
+    }
+  }
+
+  const Shape& GetInputShape(size_t indice) {
```
Comment (RandySheriffH) [inline include/onnxruntime/core/session/onnxruntime_lite_custom_op.h:996]: >GetInputShape [](http://example.com/codeflow?start=15&length=13)

make it a const function.

File: include/onnxruntime/core/session/onnxruntime_lite_custom_op.h [hunk 2]
```diff
@@ -928,6 +930,113 @@ struct OrtLiteCustomFunc : public OrtLiteCustomOp {
...
+      if (dim.IsInt()) {
+        integer_dims.push_back(dim.IsInt());
+        symbolic_dims.push_back("");
+      } else {
+        integer_dims.push_back(-1);
+        symbolic_dims.push_back(dim.AsSym());
+      }
+    }
+
+    ThrowOnError(ort_api_->SetDimensions(info, integer_dims.data(), integer_dims.size()));
```
Comment (RandySheriffH) [inline include/onnxruntime/core/session/onnxruntime_lite_custom_op.h:1027]: >ThrowOnError [](http://example.com/codeflow?start=4&length=12)

Do not throw, just return status on error.

File: include/onnxruntime/core/session/onnxruntime_lite_custom_op.h [hunk 3]
```diff
@@ -959,12 +1068,13 @@ struct OrtLiteCustomStruct : public OrtLiteCustomOp {
     size_t num_output_{};
     std::unique_ptr<CustomOp> custom_op_;
     std::string ep_{};
+    const OrtApi* ort_api_{};
```
Comment (RandySheriffH) [inline include/onnxruntime/core/session/onnxruntime_lite_custom_op.h:1071]: > const OrtApi* ort_api_{}; [](http://example.com/codeflow?start=3&length=26)

remove this.

File: include/onnxruntime/core/session/onnxruntime_lite_custom_op.h [hunk 4]
```diff
@@ -973,16 +1083,19 @@ struct OrtLiteCustomStruct : public OrtLiteCustomOp {
       kernel->custom_op_ = std::make_unique<CustomOp>(ort_api, info);
       auto self = static_cast<const OrtLiteCustomStruct*>(this_);
       kernel->ep_ = self->execution_provider_;
+      kernel->ort_api_ = ort_api;
```
Comment (RandySheriffH) [inline include/onnxruntime/core/session/onnxruntime_lite_custom_op.h:1086]: > kernel->ort_api_ = ort_api; [](http://example.com/codeflow?start=5&length=28)

remove this.

File: onnxruntime/core/graph/graph.cc [hunk 5]
```diff
@@ -29,6 +29,11 @@
 #include "core/graph/op.h"
 #include "core/graph/runtime_optimization_record_container.h"
 #include "core/graph/function_utils.h"
+#include "core/framework/error_code_helper.h"
+#include "core/session/ort_apis.h"
+#include "core/framework/tensor_type_and_shape.h"
+//#include "onnx/defs/shape_inference.h"
```
Comment (RandySheriffH) [inline onnxruntime/core/graph/graph.cc:35]: remove.

File: onnxruntime/core/graph/graph.cc [hunk 6]
```diff
@@ -4493,3 +4498,101 @@ common::Status Graph::LoadFromOrtFormat(const onnxruntime::fbs::Graph& fbs_graph
 }
 
 }  // namespace onnxruntime
+
+/*
+
+struct OrtShapeInferContext {
+ public:
```
Comment (RandySheriffH) [inline onnxruntime/core/graph/graph.cc:4505]: remove all.

File: onnxruntime/core/graph/graph.cc [hunk 7]
```diff
@@ -29,6 +29,11 @@
 #include "core/graph/op.h"
 #include "core/graph/runtime_optimization_record_container.h"
 #include "core/graph/function_utils.h"
+#include "core/framework/error_code_helper.h"
+#include "core/session/ort_apis.h"
+#include "core/framework/tensor_type_and_shape.h"
+//#include "onnx/defs/shape_inference.h"
+#include "core/framework/tensorprotoutils.h"
```
Comment (RandySheriffH) [inline onnxruntime/core/graph/graph.cc:36]: remove all.

File: onnxruntime/core/session/standalone_op_invoker.cc [hunk 8]
```diff
@@ -77,6 +77,17 @@ ORT_API_STATUS_IMPL(OrtApis::CopyKernelInfo,
...
+//                    _In_ OrtOpAttr*,
+//                    _In_ const char*,
+//                    _In_ OrtOpAttrType,
+//                    _Inout_ void*,
+//                    _In_ int) {
+//  API_IMPL_BEGIN
+//  return nullptr;
+//  API_IMPL_END
+//}
+
```
Comment (RandySheriffH) [inline onnxruntime/core/session/standalone_op_invoker.cc:90]: will remove.

File: onnxruntime/core/session/standalone_op_invoker.cc [hunk 9]
```diff
@@ -555,4 +566,15 @@ ORT_API(void, OrtApis::ReleaseKernelInfo, _Frees_ptr_opt_ OrtKernelInfo* info_co
...
+//ORT_API_STATUS_IMPL(OrtApis::ReadOpAttr,
+//                    _In_ OrtOpAttr* /*op_attr*/,
+//                    _In_ const char* /*attr_name*/,
+//                    _In_ OrtOpAttrType /*type*/,
+//                    _Inout_ void* /*data*/,
+//                    _In_ int /*len*/) {
+//  API_IMPL_BEGIN
+//  return nullptr;
+//  API_IMPL_END
+//}
```
Comment (RandySheriffH) [inline onnxruntime/core/session/standalone_op_invoker.cc:578]: will remove.

File: onnxruntime/test/testdata/custom_op_library/cpu/cpu_ops.cc [hunk 10]
```diff
@@ -206,8 +230,75 @@ Ort::Status CopyTensorArrayCombined(const Ort::Custom::Tensor<float>& first_inpu
...
+}
+
+Ort::Status AttrTesterStringShapeInfer(Ort::ShapeInferContext& ctx) {
+  CUSTOM_ENFORCE(ctx.GetAttr<std::string>("a_string") == "iamastring", "string attr mismatch");
+  std::vector<std::string> strings{"more", "strings"};
+  CUSTOM_ENFORCE(ctx.GetAttr<std::vector<std::string>>("strings") == strings, "strings attr mismatch");
+  Ort::ShapeInferContext::Shape shape5 = {{5}};
+  ctx.SetOutputShape(0, shape5);
+  return Ort::Status{nullptr};
+}
```
Comment (RandySheriffH) [inline onnxruntime/test/testdata/custom_op_library/cpu/cpu_ops.cc:262]: will remove.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 11]
```diff
@@ -4438,6 +4439,73 @@
...
+  /**
+   * Get attribute from OrtShapeInferContext
+   *
+   * \param[in] context
+   * \param[in] attr_name
+   * \param[out] attr
+   *
+   * \since Version 1.17.
+   */
+  ORT_API2_STATUS(ShapeInferContext_GetAttribute, _In_ const OrtShapeInferContext* context, _In_ const char* attr_name, _Outptr_ const OrtOpAttr** attr);
```
Comment (souptc) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4473]: >ShapeInferContext_GetAttribute [](http://example.com/codeflow?start=18&length=30)

Here we collapse the attribute from the node into ShapeInferenceContext. Maybe should doc that the ShapeInferenceContext is per-node?

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h [hunk 12]
```diff
@@ -1883,4 +1892,154 @@
...
+  for (size_t ith_input = 0; ith_input < input_count; ++ith_input) {
+    OrtTensorTypeAndShapeInfo* info{};
+    Ort::ThrowOnError(ort_api_->ShapeInferContext_GetInputTypeShape(ctx, ith_input, &info));
+    TensorTypeAndShapeInfo type_shape_info(info);
+    auto integer_shape = type_shape_info.GetShape();
+    std::vector<const char*> symbolic_shape(integer_shape.size(), {});
+    type_shape_info.GetSymbolicDimensions(&symbolic_shape[0], integer_shape.size());
+    Shape shape;
+    for (size_t ith = 0; ith < integer_shape.size(); ++ith) {
+      if (integer_shape[ith] == -1) {
```
Comment (souptc) [inline include/onnxruntime/core/session/onnxruntime_cxx_inline.h:1908]: >if (integer_shape[ith] == -1)  [](http://example.com/codeflow?start=6&length=30)

i remember onnx use "-1" as special shape in op like Reshape. use -1 to check seem not safe.

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h [hunk 13]
```diff
@@ -1883,4 +1892,154 @@
...
+
+  std::vector<int64_t> integer_dims;
+  std::vector<const char*> symbolic_dims;
+
+  for (const auto dim : shape) {
+    if (dim.IsInt()) {
+      integer_dims.push_back(dim.IsInt());
+      symbolic_dims.push_back("");
+    } else {
+      integer_dims.push_back(-1);
```
Comment (souptc) [inline include/onnxruntime/core/session/onnxruntime_cxx_inline.h:1940]: >-1 [](http://example.com/codeflow?start=29&length=2)

similar concern as above comment

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 12 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h]: Removed.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 11 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4473]: Done.

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 13 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h]: Switched to a different value - we do not use integer under the condition of symbol whatsowever.

File: include/onnxruntime/core/session/onnxruntime_cxx_api.h [hunk 14]
```diff
@@ -2156,6 +2156,75 @@
               size_t output_count);
 };
 
+struct ShapeInferContext {
+  struct SymbolicInteger {
```
Comment (souptc) [inline include/onnxruntime/core/session/onnxruntime_cxx_api.h:2160]: please add doc string as it is a new type in public api

**Review by souptc — DISMISSED** [review]
:shipit:

**Review by souptc — APPROVED** [review]

**Pull request merged**